src/sbvr-api/abstract-sql.ts

```diff
@@ -0,0 +1,188 @@
+import * as Promise from 'bluebird';
+import * as _ from 'lodash';
+
+import * as env from '../config-loader/env';
+import memoizeWeak = require('memoizee/weak');
+import { ODataBinds, odataNameToSqlName } from '@resin/odata-to-abstract-sql';
+import { SqlCompilationError } from './errors';
+import deepFreeze = require('deep-freeze');
+import { ODataRequest } from './uri-parser';
+import * as sbvrUtils from './sbvr-utils';
+import * as AbstractSQLCompiler from '@resin/abstract-sql-compiler';
+
+const getMemoizedCompileRule = (engine: AbstractSQLCompiler.Engines) =>
+	memoizeWeak(
+		(abstractSqlQuery: AbstractSQLCompiler.AbstractSqlQuery) => {
+			const sqlQuery = AbstractSQLCompiler[engine].compileRule(
+				abstractSqlQuery,
+			);
+			const modifiedFields = AbstractSQLCompiler[engine].getModifiedFields(
+				abstractSqlQuery,
+			);
+			if (modifiedFields != null) {
+				deepFreeze(modifiedFields);
+			}
+			return {
+				sqlQuery,
+				modifiedFields,
+			};
+		},
+		{ max: env.cache.abstractSqlCompiler.max },
+	);
+
+export const compileRequest = (request: ODataRequest) => {
+	if (request.abstractSqlQuery != null) {
```

Comment:
If I do that I'll have to `return request` in two spots, the current style is to make it clearer that we *always* `return request` regardless of whether we modify it or not.  Introducing an early return means you have to check all the returns and realise that they match/always return the same object that was passed in.  I can change it but that is the reasoning why it's not an early return and the trade-off in readability

Reply:
I noticed and I don't mind either way, but I personally still think it's better since the bulk of the function is in that `if` branch. Annotating the function with the return type would offer the guarantee of returning the `request` (and also be nice to readers of the code).